File: zbestcomic/src/main/java/vn/loitp/app/activity/home/allmanga/realm/RealmController.java
package vn.loitp.app.activity.home.allmanga.realm;

import android.app.Activity;
import android.app.Application;
import android.support.v4.app.Fragment;

import io.realm.Realm;
import io.realm.RealmResults;
import vn.loitp.app.model.comic.Comic;

public class RealmController {

    private static RealmController instance;
    private final Realm realm;

    public RealmController(Application application) {
        realm = Realm.getDefaultInstance();
    }

    public static RealmController with(Fragment fragment) {
        if (instance == null) {
            instance = new RealmController(fragment.getActivity().getApplication());
        }
        return instance;
    }

    public static RealmController with(Activity activity) {
        if (instance == null) {
            instance = new RealmController(activity.getApplication());
        }
        return instance;
    }

    public static RealmController with(Application application) {
        if (instance == null) {
            instance = new RealmController(application);
        }
        return instance;
    }

    public static RealmController getInstance() {
        return instance;
    }

    public Realm getRealm() {
        return realm;
    }

    //Refresh the realm istance
    public void refresh() {
        realm.refresh();
    }

    //clear all objects from Book.class
    public void clearAll() {
        realm.beginTransaction();
        realm.clear(Comic.class);
        realm.commitTransaction();
    }

    //find all objects in the Book.class
    public RealmResults<Comic> getBooks() {
        return realm.where(Comic.class).findAll();
    }

    //query a single item with the given id
    public Comic getBook(String id) {
        return realm.where(Comic.class).equalTo("id", id).findFirst();
    }

    //check if Book.class is empty
    public boolean hasBooks() {
        return !realm.allObjects(Comic.class).isEmpty();
    }

    //query example
    public RealmResults<Comic> queryedBooks() {
        return realm.where(Comic.class)
                .contains("author", "Author 0")
                .or()
                .contains("title", "Realm")
                .findAll();

    }
}

File: app/src/main/java/vn/loitp/app/a/database/realm/RealmController.kt
package vn.loitp.app.a.database.realm

import android.app.Activity
import android.app.Application
import androidx.fragment.app.Fragment
import io.realm.Realm
import io.realm.RealmResults
import java.util.*

class RealmController {
    val realm: Realm = Realm.getDefaultInstance()

    val myBookList: RealmResults<MyBook>
        get() = realm.where(MyBook::class.java).findAll()

    val myBookListSortByID: List<MyBook>
        get() {
            val myBookList = ArrayList(myBookList)
            sort(myBookList)
            return myBookList
        }

    fun refresh() {
        realm.refresh()
    }

    fun clearAllMyBook() {
        realm.beginTransaction()
        realm.clear(MyBook::class.java)
        realm.commitTransaction()
    }

    fun getMyBook(id: Long): MyBook {
        return realm.where(MyBook::class.java).equalTo(FIELD_ID, id).findFirst()
    }

    fun getMyBook(myBook: MyBook): RealmResults<MyBook> {
        return realm.where(MyBook::class.java)
            .equalTo(FIELD_TITLE, myBook.title)
            .equalTo(FIELD_ID, myBook.id)
            .findAll()
    }

    private fun sort(myBookList: List<MyBook>) {
        Collections.sort(myBookList) { obj1, obj2 ->
            // ## Ascending order
            // return obj1.firstName.compareToIgnoreCase(obj2.firstName);
            java.lang.Long.valueOf(obj1.id).compareTo(obj2.id)

            // ## Descending order
            // return obj2.firstName.compareToIgnoreCase(obj1.firstName);
            // return Integer.valueOf(obj2.empId).compareTo(obj1.empId);
        }
    }

    @Suppress("unused")
    fun hasMyBook(): Boolean {
        return !realm.allObjects(MyBook::class.java).isEmpty()
    }

    @Suppress("unused")
    fun hasMyBooks(): RealmResults<MyBook> {
        return realm.where(MyBook::class.java)
            .contains(FIELD_AUTHOR, "Author 0")
            .or()
            .contains(FIELD_TITLE, "Realm")
            .findAll()
    }

    companion object {
        const val FIELD_ID = "id"
        const val FIELD_TITLE = "title"
        const val FIELD_AUTHOR = "author"
        private var ins: RealmController? = null

        fun getInstance(activity: Activity): RealmController {
            if (ins == null) {
                ins = RealmController()
            }
            return ins!!
        }

        fun with(fragment: Fragment): RealmController {
            if (ins == null) {
                ins = RealmController()
            }
            return ins!!
        }

        fun with(activity: Activity): RealmController {
            if (ins == null) {
                ins = RealmController()
            }
            return ins!!
        }

        fun with(application: Application): RealmController {
            if (ins == null) {
                ins = RealmController()
            }
            return ins!!
        }
    }
}
